### PR TITLE
Improve UX when falling back between Display Servers

### PR DESCRIPTION
**Pull request by deralmas**
The current UX for DS fallbacks is a bit confusing, as there's quite a bit of (possibly unrelated) errors and then a working game:

<details><summary>Example log from X11 to Wayland</summary>

```
$ DISPLAY=/dev/null bin/godot.linuxbsd.editor.dev.x86_64 --display-driver x11 --rendering-driver opengl3
Godot Engine v4.3.dev.custom_build.1313cd90a (2024-05-09 09:51:35 UTC) - https://godotengine.org
Error loading shared library libXinerama.so.1: No such file or directory
ERROR: X11 Display is not available
   at: DisplayServerX11 (platform/linuxbsd/x11/display_server_x11.cpp:5861)
Your video card drivers seem not to support the required OpenGL 3.3 version.

If possible, consider updating your video card drivers.

If you recently updated your video card drivers, try rebooting.
Plugin "GTK3 plugin" uses conflicting symbol "png_free".
Error loading shared library libspeechd.so.2: No such file or directory
Invalid value (0) for DRI_PRIME. Should be > 0
Inconsistent value (1) for DRI_PRIME. Should be < 1 (GPU devices count). Using: 0
Inconsistent value (2) for DRI_PRIME. Should be < 1 (GPU devices count). Using: 0
Inconsistent value (3) for DRI_PRIME. Should be < 1 (GPU devices count). Using: 0
OpenGL API 4.6 (Core Profile) Mesa 23.3.5 - Compatibility - Using Device: AMD - AMD Radeon Vega 8 Graphics (radeonsi, raven, LLVM 17.0.6, DRM 3.57, 6.8.8)
```

</details>

This patchset[^1] aims to improve the situation where possible, without introducing any big change yet:

 - From core, we display a warning whenever a fallback is being attempted;

 - From X11, we alert the user about a failed rendering driver initialization only if it was actually attempted;

 - From Wayland we actually implement the aforementioned alerts, as they were missing (oops).

<details><summary>New log when falling back from X11 to Wayland</summary>

```
$ DISPLAY=/dev/null bin/godot.linuxbsd.editor.dev.x86_64 --display-driver x11 --rendering-driver opengl3
Godot Engine v4.3.dev.custom_build.1313cd90a (2024-05-09 09:51:35 UTC) - https://godotengine.org
Error loading shared library libXinerama.so.1: No such file or directory
ERROR: X11 Display is not available
   at: DisplayServerX11 (platform/linuxbsd/x11/display_server_x11.cpp:5842)
WARNING: Display driver x11 failed, falling back to wayland.
     at: setup2 (main/main.cpp:2701)
Plugin "GTK3 plugin" uses conflicting symbol "png_free".
Error loading shared library libspeechd.so.2: No such file or directory
Invalid value (0) for DRI_PRIME. Should be > 0
Inconsistent value (1) for DRI_PRIME. Should be < 1 (GPU devices count). Using: 0
Inconsistent value (2) for DRI_PRIME. Should be < 1 (GPU devices count). Using: 0
Inconsistent value (3) for DRI_PRIME. Should be < 1 (GPU devices count). Using: 0
OpenGL API 4.6 (Core Profile) Mesa 23.3.5 - Compatibility - Using Device: AMD - AMD Radeon Vega 8 Graphics (radeonsi, raven, LLVM 17.0.6, DRM 3.57, 6.8.8)
```

</details>

It's not perfect by any means (the DRI_PRIME stuff is especially annoying), but those were some fairly low-hanging fruits and so this will do in the meantime.

[^1]: Yes, sorry for making _three_ commits but they feel all equally sensitive and unrelated. Please tell me if a single squashed commit is preferred.

### CODE REVIEW

File: platform/linuxbsd/x11/display_server_x11.cpp [hunk 1]
```diff
@@ -6062,10 +6043,24 @@ DisplayServerX11::DisplayServerX11(const String &p_rendering_driver, WindowMode
...
+								"If you recently updated your video card drivers, try rebooting.",
+								executable_name),
+						"Unable to initialize Vulkan video driver");
+			}
+
 			return;
 		}
 		driver_found = true;
 	}
+
```
Comment (akien-mga) [inline platform/linuxbsd/x11/display_server_x11.cpp:6063]: Nit, but I wouldn't add a line here, the `#endif` should stay close to the scope it related to. You can add one after though if you feel it's too crammed.

File: platform/linuxbsd/x11/display_server_x11.cpp [hunk 2]
```diff
@@ -6136,7 +6138,14 @@ DisplayServerX11::DisplayServerX11(const String &p_rendering_driver, WindowMode
 			memdelete(gl_manager_egl);
 			gl_manager_egl = nullptr;
 			r_error = ERR_UNAVAILABLE;
-			ERR_FAIL_MSG("Could not initialize OpenGLES.");
+
+			OS::get_singleton()->alert(
+					"Your video card drivers seem not to support the required OpenGL ES 3 version.\n\n"
```
Comment (akien-mga) [inline platform/linuxbsd/x11/display_server_x11.cpp:6143]:
```suggestion
					"Your video card drivers seem not to support the required OpenGL ES 3.0 version.\n\n"
```

File: platform/linuxbsd/x11/display_server_x11.cpp [hunk 3]
```diff
@@ -6062,10 +6043,24 @@ DisplayServerX11::DisplayServerX11(const String &p_rendering_driver, WindowMode
...
+						vformat("Your video card drivers seem not to support the required Vulkan version.\n\n"
+								"If possible, consider updating your video card drivers or using the OpenGL 3 driver.\n\n"
+								"You can enable the OpenGL 3 driver by starting the engine from the\n"
+								"command line with the command:\n\n    \"%s\" --rendering-driver opengl3\n\n"
+								"If you recently updated your video card drivers, try rebooting.",
+								executable_name),
+						"Unable to initialize Vulkan video driver");
+			}
+
 			return;
```
Comment (akien-mga) [inline platform/linuxbsd/x11/display_server_x11.cpp:6059]: I notice that for other drivers we `ERR_FAIL_MSG`, but for Vulkan we print an error at the start with slightly different formatting, and then return. Should we unify those?

The same comment applies to Wayland if we do.

File: platform/linuxbsd/wayland/display_server_wayland.cpp [hunk 4]
```diff
@@ -1355,12 +1373,25 @@ DisplayServerWayland::DisplayServerWayland(const String &p_rendering_driver, Win
 				memdelete(egl_manager);
 				egl_manager = nullptr;
 				r_error = ERR_CANT_CREATE;
-				ERR_FAIL_MSG("Could not initialize GLES3.");
+
+				OS::get_singleton()->alert(
+						"Your video card drivers seem not to support the required OpenGL ES 3 version.\n\n"
```
Comment (akien-mga) [inline platform/linuxbsd/wayland/display_server_wayland.cpp:1378]:
```suggestion
						"Your video card drivers seem not to support the required OpenGL ES 3.0 version.\n\n"
```

File: platform/linuxbsd/x11/display_server_x11.cpp (hunk 1 above)
Comment (bruvzg) [reply platform/linuxbsd/x11/display_server_x11.cpp:6063]: I would add `// RD_ENABLED` to this endif (and maybe do the same to the rest of renderer init ifdefs), it's nested and with the added message it's a pretty large block.

**Review by bruvzg — APPROVED** [review]
Seems fine.

Handling fallbacks and messages seems to be a bit different in the different platforms, it probably should be unified at some point.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**